apps/module-presale/contracts/Presale.sol

```diff
@@ -0,0 +1,362 @@
+        );
+        totalDaiRaised = totalDaiRaised.add(_daiToSpend);
+
+        // Keep track of dai spent in this purchase for later refunding.
+        purchases[msg.sender][purchaseId] = _daiToSpend;
+
+        emit TokensPurchased(msg.sender, _daiToSpend, tokensToSell, purchaseId);
+    }
+
+    /**
+    * @notice Refunds a purchase made by `_buyer`, with id `_purchaseId`. Each purchase has a purchase id, and needs to be refunded separately.
+    * @param _buyer address The buyer address to refund.
+    * @param _purchaseId uint256 The purchase id to refund.
+    */
+    function refund(address _buyer, uint256 _purchaseId) public {
+        require(currentSaleState() == SaleState.Refunding, ERROR_INVALID_STATE);
+
+        // Recall how much dai to refund for this purchase.
+        uint256 daiToRefund = purchases[_buyer][_purchaseId];
+        require(daiToRefund > 0, ERROR_NOTHING_TO_REFUND);
+        purchases[_buyer][_purchaseId] = 0;
+
+        // (presale) ~~~> dai ~~~> (buyer)
+        require(daiToken.transfer(_buyer, daiToRefund), ERROR_DAI_TRANSFER_REVERTED);
+
+        // (buyer) ~~~> project tokens ~~~> (Token manager)
+        // Note: this assumes that the buyer didn't transfer any of the vested tokens.
+        // The assumption can be made, considering the imposed restriction of fundingPeriod < vestingCliffPeriod < vestingCompletePeriod.
+        (uint256 tokensSold,,,,,) = projectTokenManager.getVesting(_buyer, _purchaseId);
+        projectTokenManager.revokeVesting(_buyer, _purchaseId);
+
+        // (Token manager) ~~~> project tokens ~~~> (burn 💥)
+        projectTokenManager.burn(address(projectTokenManager), tokensSold);
+
+        emit TokensRefunded(_buyer, daiToRefund, tokensSold, _purchaseId);
+    }
+
+    /**
+    * @notice Closes a sale that has reached the funding goal, sending raised funds to the fundraising pool and the beneficiary address, and initializes the Fundraising app by adding the raised funds as collateral tokens.
+    */
+    function close() public {
+        require(currentSaleState() == SaleState.GoalReached, ERROR_INVALID_STATE);
+
+        // (presale) ~~~> dai ~~~> (beneficiary)
+        uint256 daiForBeneficiary = totalDaiRaised.mul(percentFundingForBeneficiary).div(PPM);
+        require(daiToken.transfer(beneficiaryAddress, daiForBeneficiary), ERROR_DAI_TRANSFER_REVERTED);
+
+        // (presale) ~~~> dai ~~~> (pool)
+        uint256 daiForPool = daiToken.balanceOf(address(this));
+        require(daiToken.transfer(fundraisingPool, daiForPool), ERROR_DAI_TRANSFER_REVERTED);
+
+        // Initialize Fundraising app.
+        fundraisingController.addCollateralToken(
+            daiToken,
+            0, /* virtualSupply */
+            0, /* virtualBalance */
+            uint32(CONNECTOR_WEIGHT_PPM), /* reserveRatio */
+            3 * 10 ** 18, /* slippage */
+            tapRate,
+            0 /* floor */
+        );
+        fundraisingInitialized = true;
```

Comment:
I think we should also add a function in the `AragonFundraisingController` to tell it that the presale is done. Based on this information the controller could enable or disable the `openBuyOrder` and `openSellOrder` methods.

Reply:
Interesting. This path would also greatly simplify the Presale app, since it would not need to know anything about the Fundraising app, or other Fundraising modules for that matter (instead of a Pool it could just expect a pool address). A function in the Fundraising app could be externally called later to unlock it, which would simply check if the state of the Presale app to be Closed.

Reply:
I think that’s the best way indeed. Can you work on updating the implementation that way ?

Reply:
@osarrouy done. Notice that Presale.sol no longer knows nothing about other fundraising modules. When the sale closes, it simply changes its state to "Closed" and sends the collected funds to the `beneficiaryAddress` and the `fundraisingPool` addresses. The fundraising controller's `addCollateraltoken(...)` function is no longer called when closing the presale.

Also, notice that I previously had to deploy a real fundraising controller and other fundraising modules to be able to call the `addCollateraltoken(...)` function. Despite the Presale contract no longer having to know about these, the code associated with their deployment has not been removed from the tests, only commented out (see [test/common/deploy.js' prepareDefaultSetup(...)](https://github.com/AragonBlack/fundraising/blob/59504f80d1b9b7ebef2b9e641752caa7e8b5efa9/apps/module-presale/test/common/deploy.js#L237)). I figured that you will be using this code when you hook up the presale with the Fundraising controller (in reverse order this time).